Add bracket colors plugin

asifamin13 opened the pull request:
# Bracketcolors plugin

- Color { }, [ ], ( ) based on nesting order to make it easier to see the start/end bracket
- Ignore brackets in non source code (strings, comments, doc strings, etc.)
- Language agnostic 



## How it works 

Allocates 3 indicators (starting at `INDICATOR_IME` - 3) to color each bracket pair as determined by `SCI_BRACEMATCH`. Ignore non source brackets as determined by `highlighting_is_code_style`

## Dependencies

- Geany 1.38
  - Need `INDICATOR_IME` from updated scintilla release
- C++17
  - Mainly for `std::map` 

### Demo 
<img width="290" alt="image" src="https://user-images.githubusercontent.com/122637911/215630933-08786dc1-c903-4045-b4aa-64cc08d32b8f.png">

rdipardo commented:
Perhaps the description and/or commit message should mention #1148?

asifamin13 commented:
I updated the commit message with a reference to that issue. Raindow brackets sounds cool, I wish I knew of that issue before starting this

elextr commented:
While "rainbow colours" sounds nice "Bracketcolors" is more descriptive, so a user looking for it is more likely to find it.

eht16 commented on bracketcolors/README:
```diff
@@ -0,0 +1,32 @@
+Color brackets, parenthesis, and braces
+===================
```
RestructuredText (the text format used here) requires that underlines are exactly as long as the heading above.

asifamin13 replied:
thanks, fixed!

eht16 commented on bracketcolors/src/BracketMap.cc:
```diff
@@ -0,0 +1,143 @@
+/*
+ *      BracketMap.cc
+ *
+ *      Copyright 2013 Asif Amin <asifamin@utexas.edu>
```
2013? I guess it's a typo :).
If so, ideally change all occurrences.

asifamin13 replied:
thanks for catching this, fixed!

eht16 commented on build/ax_cxx_compile_stdcxx_17.m4:
```diff
@@ -0,0 +1,35 @@
+# =============================================================================
```
Are these two "ax_cxx_*" M4 macros copied from Geany?

asifamin13 replied:
yes, is there a better way to handle this? I'm new to the autotools world, so any recommendations would be appreciated

eht16 replied:
I don't know if there is a better way :(. Asked primarily for curiosity.

eht16 commented:
Looks good! Thanks.

> While "rainbow colours" sounds nice "Bracketcolors" is more descriptive, so a user looking for it is more likely to find it.

Full agree.
@asifamin13 maybe you could just mention "rainbow colors" in the README, that might already suffice that users can find it also with this name.

I did a quick code review without having a deeper look at the logic.
From a quick test, it works as advertised and I got many fancy brackets :).

A few comments @asifamin13:
- Could you add the plugin name in https://github.com/geany/geany-plugins/blob/master/build/geany-plugins.nsi#L176 (this is for the Windows installer, use `bracketcolors.dll`)
- As Geany itself also colorizes the brackets in an idle callback, did you check that both implementations do not interfere with each other?
- I got a few compiler warnings which might be worth looking into:
```
BracketMap.cc: In member function 'void BracketMap::Show()':
BracketMap.cc:122:21: warning: loop variable 'it' creates a copy from type 'const std::pair<const int, std::tuple<int, int> >' [-Wrange-loop-construct]
  122 |     for (const auto it : mBracketMap) {
      |                     ^~
BracketMap.cc:122:21: note: use reference type to prevent copying
  122 |     for (const auto it : mBracketMap) {
      |                     ^~
      |                     &
bracketcolors.cc: In function 'gboolean utils_parse_color(const gchar*, GdkColor*)':
bracketcolors.cc:276:27: warning: 'gboolean gdk_color_parse(const gchar*, GdkColor*)' is deprecated: Use 'gdk_rgba_parse' instead [-Wdeprecated-declarations]
  276 |     return gdk_color_parse(spec, color);
      |            ~~~~~~~~~~~~~~~^~~~~~~~~~~~~
In file included from /usr/include/gtk-3.0/gdk/gdkcairo.h:26,
                 from /usr/include/gtk-3.0/gdk/gdk.h:33,
                 from /usr/include/gtk-3.0/gtk/gtk.h:30,
                 from /home/enrico/apps/include/geany/gtkcompat.h:30,
                 from /home/enrico/apps/include/geany/editor.h:27,
                 from /home/enrico/apps/include/geany/document.h:31,
                 from /home/enrico/apps/include/geany/build.h:26,
                 from /home/enrico/apps/include/geany/geanyplugin.h:36,
                 from bracketcolors.cc:40:
/usr/include/gtk-3.0/gdk/deprecated/gdkcolor.h:79:11: note: declared here
   79 | gboolean  gdk_color_parse     (const gchar    *spec,
      |           ^~~~~~~~~~~~~~~
```

eht16 commented on bracketcolors/src/BracketMap.h:
```diff
@@ -0,0 +1,76 @@
+        return std::get<1>(bracket);
+    }
+};
+
```
Why so many empty lines?

asifamin13 replied:
this was from a standard template I use, fixed

eht16 reviewed (changes requested):
Oh, and could you please add your plugin to the list in the top README?

eht16 commented on bracketcolors/src/BracketMap.cc:
```diff
@@ -0,0 +1,143 @@
+/*
+
+----------------------------------------------------------------------------- */
+{   g_debug("%s: Showing bracket map ...", __FUNCTION__);
```
Should those `g_debug` statements remain in the code? Also the commented ones?

asifamin13 replied:
they aren't needed anymore. I removed the commented ones as well

asifamin13 commented:
> 



> Looks good! Thanks.
> 
> > While "rainbow colours" sounds nice "Bracketcolors" is more descriptive, so a user looking for it is more likely to find it.
> 
> Full agree. @asifamin13 maybe you could just mention "rainbow colors" in the README, that might already suffice that users can find it also with this name.
> 
> I did a quick code review without having a deeper look at the logic. From a quick test, it works as advertised and I got many fancy brackets :).
> 
> A few comments @asifamin13:
> 
> * Could you add the plugin name in https://github.com/geany/geany-plugins/blob/master/build/geany-plugins.nsi#L176 (this is for the Windows installer, use `bracketcolors.dll`)
> * As Geany itself also colorizes the brackets in an idle callback, did you check that both implementations do not interfere with each other?
> * I got a few compiler warnings which might be worth looking into:
> 
> ```
> BracketMap.cc: In member function 'void BracketMap::Show()':
> BracketMap.cc:122:21: warning: loop variable 'it' creates a copy from type 'const std::pair<const int, std::tuple<int, int> >' [-Wrange-loop-construct]
>   122 |     for (const auto it : mBracketMap) {
>       |                     ^~
> BracketMap.cc:122:21: note: use reference type to prevent copying
>   122 |     for (const auto it : mBracketMap) {
>       |                     ^~
>       |                     &
> bracketcolors.cc: In function 'gboolean utils_parse_color(const gchar*, GdkColor*)':
> bracketcolors.cc:276:27: warning: 'gboolean gdk_color_parse(const gchar*, GdkColor*)' is deprecated: Use 'gdk_rgba_parse' instead [-Wdeprecated-declarations]
>   276 |     return gdk_color_parse(spec, color);
>       |            ~~~~~~~~~~~~~~~^~~~~~~~~~~~~
> In file included from /usr/include/gtk-3.0/gdk/gdkcairo.h:26,
>                  from /usr/include/gtk-3.0/gdk/gdk.h:33,
>                  from /usr/include/gtk-3.0/gtk/gtk.h:30,
>                  from /home/enrico/apps/include/geany/gtkcompat.h:30,
>                  from /home/enrico/apps/include/geany/editor.h:27,
>                  from /home/enrico/apps/include/geany/document.h:31,
>                  from /home/enrico/apps/include/geany/build.h:26,
>                  from /home/enrico/apps/include/geany/geanyplugin.h:36,
>                  from bracketcolors.cc:40:
> /usr/include/gtk-3.0/gdk/deprecated/gdkcolor.h:79:11: note: declared here
>    79 | gboolean  gdk_color_parse     (const gchar    *spec,
>       |           ^~~~~~~~~~~~~~~
> ```

Thanks for the feedback! 

- The deprecated `gdk*` warning are ok. I removed the unused debugging code that was causing the others. 
- I added the bracketcolors entry to `geany-plugins.nsi`. 
- I confirmed my plugin behaves correctly when Geany colors brackets itself, like when brackets are missing

asifamin13 commented:
@eht16 I incorporated your suggested changes. I'm still learning Github, I apologize if I'm not navigating this pull request correctly 😅 (at my work, I'm used to Gitlab)

eht16 commented:
> @eht16 I incorporated your suggested changes. I'm still learning Github, I apologize if I'm not navigating this pull request correctly sweat_smile (at my work, I'm used to Gitlab)

No worries, it's fine.
Thanks for the changes. The PR looks good to me and I would merge it in a few days if nobody objects.

b4n commented on bracketcolors/src/bracketcolors.cc:
```diff
@@ -0,0 +1,1303 @@
+
+
+// -----------------------------------------------------------------------------
+    static gchar char_at(ScintillaObject *sci, gint pos)
+/*
+
+----------------------------------------------------------------------------- */
+{
+    return sci_get_char_at(sci, pos);
+}
```
Just curious: why the wrapper?

rdipardo replied:
> Just curious: why the wrapper?

The [documented return type of `SCI_GETCHARAT`] is `int`, presumably for Unicode characters up to 32 bits in length. Squeezing that into an 8-bit(?) `(g)char` needs an explicit cast.

The (implied) point that just using the returned `(g)int` without a cast is worth considering. At the same time, the `pos` argument might be safer as the (platform independent) [`Sci_Position` type]. The API structures that can *only* index 32-bit positions are [slated for eventual deprecation].

[documented return type of `SCI_GETCHARAT`]: https://www.scintilla.org/ScintillaDoc.html#SCI_GETCHARAT
[`Sci_Position` type]: https://github.com/geany/geany/blob/8a9e9be88545f83c91d378d6662790efa38959d4/scintilla/include/Sci_Position.h#L14-L15
[slated for eventual deprecation]: https://groups.google.com/g/scintilla-interest/c/mPLwYdC0-FE#:~:text=SCI_GETTEXTRANGE%20will%20be%20deprecated%20at%20some%20point%20(likely%206.0)

b4n replied:
> The [documented return type of `SCI_GETCHARAT`](https://www.scintilla.org/ScintillaDoc.html#SCI_GETCHARAT) is `int`, presumably for Unicode characters up to 32 bits in length.

No, that's probably either because `int` is "the" basic numerical type in C, or because of other historical reasons (and technically it is returned as a `sptr_t`). In practice, the returned value comes from `Document::ChatAt()` which returns a `char`. So if if want to be picky, it's actually a `char` stored into a `sptr_t`.

> Squeezing that into an 8-bit(?) `(g)char` needs an explicit cast.

Well, yes, but… actually no: this does not actually wrap `SCI_GETCHARAT`, but Geany's own wrapper, `sci_get_char_at()` *which has the exact same signature*.

asifamin13 replied:
When I was writing this, I was using the `autoclose` plugin as a template [which has this wrapper as well](https://github.com/geany/geany-plugins/blob/dc6980e5df24434597b37a5a7064837b3f5158e8/autoclose/src/autoclose.c#L93). I suspect the authors of `autoclose` used the wrapper for convenience since it's used a lot, I don't really need it as bad to be honest.

elextr commented:
? why removed

rdipardo commented:
> ? why removed

To get [your attention](https://github.com/geany/geany-plugins/pull/1241#issuecomment-1464731699), it seems.

elextr commented:
Nah, it happens that people learning Git can stuff it up so bad its easier to start again, no problem.

But unfortunately that disconnects the discussion.

b4n commented:
Instead of breaking the flow, just force-push your branch: `git push -f origin bracketcolors_v1`. You probably can even revive this one by simply reopening it, and if GitHub isn't smart enough, there's a bunch of tricks we could do to force it's hand.

b4n commented:
> Nah, it happens that people learning Git can stuff it up so bad its easier to start again, no problem.

Well, it usually is not easier, but people just don't know it 😅

elextr commented:
> > Nah, it happens that people learning Git can stuff it up so bad its easier to start again, no problem.

> Well, it usually is not easier, but people just don't know it :sweat_smile:

Thats learning :grin:

asifamin13 commented:
> Instead of breaking the flow, just force-push your branch: `git push -f origin bracketcolors_v1`. You probably can even revive this one by simply reopening it, and if GitHub isn't smart enough, there's a bunch of tricks we could do to force it's hand.

This is what I did, along with squashing several commits similar [to this](https://stackoverflow.com/questions/5667884/how-to-squash-commits-in-git-after-they-have-been-pushed). I made the mistake of forgetting that I pressed the "sync with upstream" button (due to a warning I saw on the UI), and I ended up including (and taking ownership) of the upstream commits. I then panicked and renamed that branch 'bracketcolors_v1_broken' and started over again with a new 'bracketcolors_v1' branch where I cherry picked the relevant commits I cared about. But that ended up closing this pull request 💀 

If any of yall can somehow reverse that mess, I owe you a beer.

b4n commented:
> If any of yall can somehow reverse that mess, I owe you a beer.

Challenge accepted, I'll see if I can convince GitHub quickly :beer: 
Otherwise don't worry, it's not the end of the world

b4n commented:
@asifamin13 actually I can't do it because I don't have permission to push to your branch.

If you want to try it out, I would think you'd have to do this:
1. Close #1241 (GitHub doesn't like 2 PRs referring to the same branch)
2. Recover this PR with `git push -f asifamin13 1e9a8b0c32d15c27ba448e689545c0bc5c9b97aa:bracketcolors_v1` (assuming the remote to your own fork is *asifamin13*)
3. Reopen this PR
4. Re-add your latest changes: `git push -f asifamin13 bracketcolors_v1`.

Do not pull in between the steps, as step 2 is restoring your old state temporarily for GitHub's sake, and step 4 is trying to re-put your new changes there.

But again, if you're feeling out of your depth, there are two options:
* allow maintainers to push to your PR
* don't worry and accept that it's the cost of learning

asifamin13 commented:
@b4n I added you as a collaborator, let me know if there are any problems.

 I had some trouble with step 2
```shell
fatal: bad object 1e9a8b0c32d15c27ba448e689545c0bc5c9b97aa
fatal: the remote end hung up unexpectedly
```
I'm a bit out of my depth but I'm willing to learn. Where does that hash come from?

b4n commented:
> @b4n I added you as a collaborator, let me know if there are any problems.

Here you go, that PR is restored and should match what you had in #1241, :beers: 
You can now push whatever you need (including force-pushing if necessary) and it should properly update this PR.

> I had some trouble with step 2
> 
> ```shell
> fatal: bad object 1e9a8b0c32d15c27ba448e689545c0bc5c9b97aa
> fatal: the remote end hung up unexpectedly
> ```
> 
> I'm a bit out of my depth but I'm willing to learn. Where does that hash come from?

1e9a8b0c32d15c27ba448e689545c0bc5c9b97aa was the commit this PR was at (as seen in the commits history before I force-pushed, and as you can see in the force-push notification above).  The idea was to restore the branch to the state it had when this PR was closed so GitHub sees it as matching this PR, and accept re-opening this PR.  If you didn't have that hash anymore, it means that Git's garbage collection got rid of it on your end (as it was an old commit not referenced by anything anymore);  you could have gotten it back by fetching it from GitHub (as I did myself, `git fetch asifamin13 1e9a8b0c32d15c27ba448e689545c0bc5c9b97aa`).

asifamin13 commented:
@b4n you are a life saver, your next beer is on me 🍻 

I didn't realize git had garbage collection, I'll be sure to `git fetch` next time

This PR is back on track now 😄

asifamin13 commented:
Fixed a bug to address a case when nested brackets weren't getting their color updated when their parent gets updated

rdipardo commented:
@asifamin13,

> Fixed a bug to address a case when nested brackets weren't getting their color updated when their parent gets updated

Looks like a good improvement.

In the future, though, I would concentrate on just getting a minimal working version merged, then touch it up later in small, traceable iterations. To paraphrase [the famous saying of Valéry], a pull request is never finished; it's abandoned.

[the famous saying of Valéry]: https://quoteinvestigator.com/2019/03/01/abandon

elextr commented:
> To paraphrase [the famous saying of Valéry](https://quoteinvestigator.com/2019/03/01/abandon), a pull request is never finished; it's abandoned.

Wise(ish) words, and its corollary, 'Every time a PR changes "somebody" has to re-check it and re-test it, and eventually "somebody" gets bored' :-)

asifamin13 commented:
> In the future, though, I would concentrate on just getting a minimal working version merged, then touch it up later in small, traceable iterations. To paraphrase [the famous saying of Valéry](https://quoteinvestigator.com/2019/03/01/abandon), a pull request is never finished; it's abandoned.

Understood. What it currently on this branch is the "minimal working version". There is one more feature I want to add (user defined colors via preferences menu/config file). I'm working on this now, hopefully to be finished this week. I'll keep that on a different commit 

> Wise(ish) words, and its corollary, 'Every time a PR changes "somebody" has to re-check it and re-test it, and eventually "somebody" gets bored' :-)

I'm fine being that "somebody" haha I've had fun writing this plugin.

Do new plugin PRs getting pulled in on the next geany-plugins release?

asifamin13 commented:
I added support for loading/saving settings via a `GKeyFile`

asifamin13 commented:
I've been daily driving this plugin for several months now. I think it's stable enough to be merged. 

The latest [CI build failure](https://github.com/geany/geany-plugins/actions/runs/4631643800) seems to be from misconfigured runners.

elextr commented:
The CI was awaiting a manual approval to run, approved, lets see.

Ideally needs someone who isn't the author to test it, nothing personal, but we all know how badly we test our own software :-)

asifamin13 commented:
Hmmm, the other CI builds seem to work without a problem. For some reason this one isn't getting picked up:

```
This request was automatically failed because there were no enabled runners online to process the request for more than 1 days.
```

> Ideally needs someone who isn't the author to test it, nothing personal, but we all know how badly we test our own software :-)

haha I understand. I'm always open to feedback if yall want to give it a try 😜

eht16 commented:
> Hmmm, the other CI builds seem to work without a problem. For some reason this one isn't getting picked up:
> 
> ```
> This request was automatically failed because there were no enabled runners online to process the request for more than 1 days.
> ```

Maybe this is related to the deprecated Ubuntu 18.04 build configuration in the workflow. Could you try to rebase your branch on current GIT master? The workflow/CI configuration has been changed a lot in the last weeks.
With a bit of luck, the CI will finally work then :).

asifamin13 commented:
@eht16 I did the rebase, verified it compiles/works as expected on my rhel 8 workstation. Shall we give the CI another shot?

eht16 commented:
> @eht16 I did the rebase, verified it compiles/works as expected on my rhel 8 workstation. Shall we give the CI another shot?

The CI is fine.
Though you added quite some new code in the last commit which should be reviewed.

Btw, what is the "#5" referring to you mentioned in the commit message?

asifamin13 commented:
It's from [the pull request](https://github.com/asifamin13/geany-plugins/pull/5) on my fork. 

I'd appreciate any feedback, I added some GUI elements to specify custom colors on the plugin preferences page.

asifamin13 commented:
Any chance this could make it into geany-plugins v2.0? 👀

eht16 commented:
I'm afraid it's too late since the translation string freeze already happened. But we should finally review and merge it after 2.0.

asifamin13 commented:
I think it would be useful and cool to colorize the indentation guides in a similar manner. Would this functionality be better as a separate plugin or would it make more sense to include it here but maybe rename the plugin?

elextr commented:
Good thing to ping it so "somebody" is reminded to try it and merge if they get a chance.  Otherwise things get lost/forgotten, if it happens that nobody has time immediately, don't leave it nearly a year until the next ping.

asifamin13 commented:
> Good thing to ping it so "somebody" is reminded to try it and merge if they get a chance. Otherwise things get lost/forgotten, if it happens that nobody has time immediately, don't leave it nearly a year until the next ping.

Understood! Would someone be so kind as to approve my latest changes now? 

With regards to the coloring the indent guides, I feel like it should be possible and I got pretty close but no dice yet. 

Using the scintilla API I can quickly and easily infer what the indentation color should be, but using `SSM(sci, SCI_STYLESETFORE, STYLE_INDENTGUIDE, color)` sets the guide color for the entire document not just an indentation block. 

Using indicators and `INDIC_BOX` I feel like I could probably implement the right half of this example but I personally think it looks ugly. 
![image](https://github.com/user-attachments/assets/e28c90af-4664-4def-b9c3-6f98fb4bd759)

asifamin13 commented:
Funnily enough as I was trying to color the indent guides I found a bug with my bracketcolors plugin. 

![image](https://github.com/user-attachments/assets/0a3a5b60-d94b-406d-8bab-3c67668b8798)

asifamin13 commented:
One last ping attempt until the new version :sweat_smile:  @eht16 @b4n @rdipardo @elextr

b4n commented on configure.ac:
```diff
@@ -13,6 +13,7 @@ AC_USE_SYSTEM_EXTENSIONS
 AC_PROG_CC
 AC_PROG_CC_C99
 AM_PROG_CC_C_O
+AC_PROG_CXX
```
couldn't you simply `AC_REQUIRE([AC_PROG_CXX])` in `GP_CHECK_BRACKETCOLORS`?

b4n commented:
> One last ping attempt until the new version 😅 @eht16 @b4n @rdipardo @elextr

I'm sorry we let you slip once again, but it's too late for 2.1 given this plugin isn't trivial.  Try and ping again shortly after release so we can try and merge this at the *start* of the cycle :sweat_smile:
